### DISCUSSION TITLE
Method for collecting lowerings

### DISCUSSION THREAD

**Discussion by abrown** [nemequ/wasm-simd-data#3] [General]
@nemequ, thanks for putting this together; it looks like a pretty useful repository. If I understand the lowering part of this (which I'm most interested in), I think it is doing:
 1. we write a SIMD instruction using a SIMDe intrinsic
 2. SIMDe converts it to something (?) compilable by Clang to machine code
 3. we look at the emitted machine code

But, this does raise some doubts: are these the lowerings I will see in real-world Wasm engines? Even if SIMDe's lowerings are the best possible case, they may not reflect the real world; also, it could be that they aren't the best possible case.

I don't say this to derail the use of SIMDe, that still seems useful, but I wonder if you would be interested in looking at the lowerings for engines:
 1. we write a SIMD instruction using a Wasm SIMD intrinsic
 2. Clang compiles this to a Wasm file
 3. we run the file through an engine, e.g. Node, and inspect the JIT-compiled code with perf

I can't say this latter approach is very easy but I've done something similar in the past and I'm wondering if you're open to this type of approach.

**Comment by nemequ** [comment]
> I think it is doing:
>
> 1. we write a SIMD instruction using a SIMDe intrinsic
> 2. SIMDe converts it to something (?) compilable by Clang to machine code
> 3. we look at the emitted machine code

I think that's right, but it's phrased a little oddly… to be clear, I would say

1. we write a C function which is just a trivial wrapper around a SIMDe function.
2. we use clang to compile that function, targeting whatever architecture and set of options want, and output assembly (we pass -S to clang).
3. we call llvm-mca on the assembly output to generate timing information (currently number of uOps and latency)

> But, this does raise some doubts: are these the lowerings I will see in real-world Wasm engines? Even if SIMDe's lowerings are the best possible case, they may not reflect the real world; also, it could be that they aren't the best possible case.

They're definitely not the best case right now.  Currently the implementations in SIMDe are basically crap; I mostly just added portable implementations with the intent to go back later and add optimized implementations, and I haven't done that yet. In theory, though, SIMDe's implementations will be improving soon, largely by stealing implementations from the WASM SIMD issue tracker (which has proposed implementations), and from other implementations (such as v8).  I'm not sure if I'm going to be doing that work or someone else, but either way I'm thinking within a few months it should be done.

Over time, I think that most of the implementations should converge on optimal lowerings, *especially* if there is a good source of information on what those are (like this repository).  If there is a better implementation somewhere I'm definitely going to want to add it to SIMDe, and I suspect other projects feel the same way.  So in the long term I think these differences will be a relatively minor concern.

You're right that it's not necessarily true that SIMDe is doing exactly the same thing as other implementations, but right now no two implementations are guaranteed to be doing exactly the same thing.  I don't know how many implementations there are right now, but it seems likely that the list will grow over time.

> I don't say this to derail the use of SIMDe, that still seems useful, but I wonder if you would be interested in looking at the lowerings for engines:

Absolutely.

During the discussion on Friday I was thinking more about going through each implementation to see what they did, or maybe just pulling the data out of the issue tracker and using that.  Since I'm planning on doing that anyways to improve SIMDe I figured I'd just put a little extra effort into documenting everything while I worked.  It didn't even occur to me that I could just use SIMDe to generate the data automatically; when I thought of it I actually laughed the idea off at first, but the more I thought about it the more sense it made.  All the tooling is there; it's easy to feed the output into llvm-mca to get timing information, it can be automated so it's easier to generate, maintain, and much less error-prone.  Basically, I think it is *by far* the easiest option.

If you look at [data/families](https://github.com/nemequ/wasm-simd-data/tree/main/data/families) you'll see there isn't really anything SIMDe-specific.  There is no reason you couldn't swap out SIMDe for some other lowering mechanism, and I'd be happy to have different options available for lowering using different engines.

After I have the code to do the lowering via SIMDe in place I plan to start working on something like the Intel Intrinsics Guide, or the Arm NEON Intrinsics Reference (which they just improved significantly last week) using this data.  There is no reason you couldn't plug alternate data into that, too.

If you or anyone else wants to work on generating similar data using a different implementation I'd love to have it live here alongside the SIMDe implementation.  If there is something you need changed in the data format I'm certainly open to the idea, and if I do end up making a reference site I'd be happy to integrate the data.  FWIW, I'd also be willing to accept data on other APIs (e.g., Rust) in addition to wasm_simd128.h and WAV.

If you can somehow get the assembly equivalent of whatever lowerings an implementation will generate into a file, we could probably also re-use a lot of the stuff I'm working on for the SIMDe implementation so llvm-mca would generate the timing data automatically.

**Reply by abrown** [reply]
Ok, I have a rather lengthy process for getting V8's lowerings out, but it is missing one final piece:
 1. Compile the lowering C files to Wasm
 2. Compile the Wasm file in NodeJs using `WebAssembly.instantiate` and observe the profiler information it produces; roughly, `perf record ... node --perf-prof --no-liftoff ...` (another option here would be to parse the logs emitted by `--print-wasm-code`)
 3. Use `perf inject ...` to create ELF images of each of the compiled functions and search through these to get the emitted code

The missing piece is that the output has the function prologue and epilogue added:

```
   7   │ 0000000000000040 <Function:test-0-turbofan>:
   8   │   40:   55                      push   %rbp
   9   │   41:   48 89 e5                mov    %rsp,%rbp
  10   │   44:   6a 0a                   pushq  $0xa
  11   │   46:   56                      push   %rsi
  12   │   47:   c5 f1 fe ca             vpaddd %xmm2,%xmm1,%xmm1
  13   │   4b:   48 8b e5                mov    %rbp,%rsp
  14   │   4e:   5d                      pop    %rbp
  15   │   4f:   c3                      retq   
  16   │   50:   90                      nop
  17   │   51:   0f 1f 00                nopl   (%rax)
  18   │   54:   00 00                   add    %al,(%rax)
```
If there was a good way get rid of these extra instructions then this could work. Any thoughts?

**Reply by nemequ** [reply]
Nice.

For the SIMDe version, I [ignore](https://github.com/nemequ/wasm-simd-data/blob/1e571aa4aa0c972d7ae2a7a1732b79e8fe2ad409/lowering/parse_mca.py#L42-L43) the [retq](https://github.com/nemequ/wasm-simd-data/blob/1e571aa4aa0c972d7ae2a7a1732b79e8fe2ad409/data/targets.yml#L11), but that's all that's necessary.

If you can figure out a way to recognize the prologue and epilogue, you could add [`LLVM_MCA_BEGIN`/`LLVM_MCA_END`](https://www.llvm.org/docs/CommandGuide/llvm-mca.html#using-markers-to-analyze-specific-code-blocks) comments, or just integrate the logic into the script which parses LLVM-MCA's data; I ignore the function level summary and only use instruction-level information anyways, so this shouldn't be a problem.  Have you looked at `--print-wasm-code`? Maybe it includes more information which you could use?

Is there a way to control the architecture v8 is targeting, or is it always tied to the CPUID (or, on Arm, probably `getauxval(AT_HWCAP)`/`getauxval(AT_HWCAP2)`)?

**Reply by abrown** [reply]
> Have you looked at `--print-wasm-code`? 

Yeah, it looks something like the following so I tend to prefer the ELF image since I can more easily disassemble it:

```
--- WebAssembly code ---
index: 0
kind: wasm function
compiler: TurboFan
Body (size = 32 = 28 + 4 padding)
Instructions (size = 20)
0x1dd511143380     0  55             push rbp
0x1dd511143381     1  4889e5         REX.W movq rbp,rsp
0x1dd511143384     4  6a0a           push 0xa
0x1dd511143386     6  56             push rsi
0x1dd511143387     7  c5f1feca       vpaddd xmm1,xmm1,xmm2
0x1dd51114338b     b  488be5         REX.W movq rsp,rbp
0x1dd51114338e     e  5d             pop rbp
0x1dd51114338f     f  c3             retl
0x1dd511143390    10  90             nop
0x1dd511143391    11  0f1f00         nop

Safepoints (size = 8)

RelocInfo (size = 0)

--- End code ---
```

I guess I could run a bunch of these and see if the prologue/epilogue patterns are the same. That doesn't seem too bad... 

> Is there a way to control the architecture v8 is targeting, or is it always tied to the CPUID

I see a `--target-arch` option but the documentation is not very promising: `--target-arch (The mksnapshot target arch. (mksnapshot only))`. I don't know for sure, but likely this works only with the host CPU. (This may not be the biggest problem because we could probably rent an instance for whatever architecture is needed?)

----

What is your sense on merging this approach? It would complicate the `generate_makefile.py`/`parse_mca.py` workflow a bit. I feel comfortable creating a PR for a tool that would take as input a Wasm file with a single-instruction function (I think you already create this somehow) and emit a list of instructions using tools like `perf`, `objdump`, `node`, etc. (a bit Linux specific). But I don't know exactly how you want to work that in to the current workflow--thoughts?

**Reply by nemequ** [reply]
> Yeah, it looks something like the following so I tend to prefer the ELF image since I can more easily disassemble it:

You're right, that doesn't look any better ☹

> I see a `--target-arch` option but the documentation is not very promising: `--target-arch (The mksnapshot target arch. (mksnapshot only))`. I don't know for sure, but likely this works only with the host CPU. (This may not be the biggest problem because we could probably rent an instance for whatever architecture is needed?)

Or we could probably use qemu.  IIRC QEMU's x86 SIMD support is pretty limited (I think it's really only through SSE4.1), but NEON and SVE are fanstastic, and POWER is pretty good.

QEMU could also help on x86; I think it's possible to disable a feature even if the host CPU supports it, so we might be able to use that to generate different versions (SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, AVX, AVX2, AVX-512VL, …) even if there is no way to control that in v8.
 
> What is your sense on merging this approach? It would complicate the `generate_makefile.py`/`parse_mca.py` workflow a bit. I feel comfortable creating a PR for a tool that would take as input a Wasm file with a single-instruction function (I think you already create this somehow) and emit a list of instructions using tools like `perf`, `objdump`, `node`, etc. (a bit Linux specific). But I don't know exactly how you want to work that in to the current workflow--thoughts?

Overall I'm open to it, but I think it would probably be better as a separate directory instead of modifying `generate_makefile.py`.  `parse_mca.py` is already a bit of a mess so you might want to fork that, too, I'm not sure.  I'm definitely open to restructuring the existing lowering directory a bit to better accommodate multiple implementations.

**Comment by abrown** [comment]
Ok, @nemequ, I started hacking away in the repository to try to add some of this. I wanted to make sure that I could run what you currently have so I tried `scripts/generate.py` but no luck: `No such file or directory: '.../wasm-simd-data/scripts/../simde/simde/wasm/simd128.h`. I think I should be retrieving SIMDe somehow but what is the expected way to do this--Git submodule? Or is that `..` supposed to be a `../..` to get at the parent of this project?

**Reply by nemequ** [reply]
You probably want to just clone a copy of simde into that directory.  I didn't want to make it a submodule because I just want to use the latest version without having to constantly add commits to update the submodule.  I should probably have a top-level Makefile to grab a copy of SIMDe if it doesn't already exist (among other things)…

That said, since I already have a checkout of simde I just symlink `simde` to `../simde` and it works well.

**Reply by nemequ** [reply]
Just added a top-level Makefile which should clear things up.  Not extensively tested, but it seems to work for me :)
